core/src/main/java/com/sequenceiq/cloudbreak/service/parcel/ParcelFilterService.java

```diff
@@ -40,23 +40,29 @@ public Set<ClouderaManagerProduct> filterParcelsByBlueprint(Long workspaceId, Lo
             LOGGER.debug("We can not identify one of the service from the blueprint so to stay on the safe side we will add every parcel");
             return parcels;
         }
-        Set<ClouderaManagerProduct> requiredParcels = filterParcels(workspaceId, stackId, parcels, serviceNamesInBlueprint);
+        Set<String> availableParcelNamesFromImage = imageReaderService.getParcelNames(workspaceId, stackId);
+        if (availableParcelNamesFromImage.isEmpty()) {
+            LOGGER.debug("There is no pre warmed parcel on the image therefore we assume that every parcel is required. Parcels: {}", parcels);
+            return parcels;
+        }
+        Set<ClouderaManagerProduct> requiredParcels = filterParcels(parcels, serviceNamesInBlueprint, availableParcelNamesFromImage);
         LOGGER.debug("The following parcels are used in CM based on blueprint: {}", requiredParcels);
         return requiredParcels;
     }
 
-    private Set<ClouderaManagerProduct> filterParcels(Long workspaceId, Long stackId, Set<ClouderaManagerProduct> parcels,
-            Set<String> requiredServicesInBlueprint) {
+    private Set<ClouderaManagerProduct> filterParcels(Set<ClouderaManagerProduct> parcels, Set<String> requiredServicesInBlueprint,
+            Set<String> availableParcelNamesFromImage) {
         Set<ClouderaManagerProduct> requiredParcels = new HashSet<>();
         Set<ClouderaManagerProduct> notAccessibleParcels = new HashSet<>();
         Iterator<ClouderaManagerProduct> parcelIterator = parcels.iterator();
+        addCustomParcelsIfPresent(availableParcelNamesFromImage, parcels, requiredParcels);
```

Comment:
it would be nicer if you could refactor the syntax this way:
`requiredParcels.addAll(collectCustomParcelsIfPresent)`
and the collect method could just return with `customParcels`